File: scripts/generate-texture.mjs
import fs from "node:fs/promises";
import path from "node:path";

const args = Object.fromEntries(
  process.argv.slice(2).map((arg) => {
    const [k, ...rest] = arg.replace(/^--/, "").split("=");
    return [k, rest.join("=") || true];
  })
);

const endpoint = (args.endpoint || process.env.COMFYUI_URL || "http://127.0.0.1:8188").replace(/\/$/, "");
const presetName = args.preset || "water";
const size = Number(args.size || 1024);
const steps = Number(args.steps || 9);
const cfg = Number(args.cfg || 1.2);
const seed = Number(args.seed || Math.floor(Math.random() * 2147483647));
const timeoutSec = Number(args.timeout || 900);
const checkpoint = args.ckpt || "zImageTurbo_turbo.safetensors";
const outputName = args.out || `${presetName}_${Date.now()}.png`;

const presetPath = path.resolve("scripts", "texture-presets.json");
const presets = JSON.parse(await fs.readFile(presetPath, "utf8"));
const preset = presets[presetName];
if (!preset) {
  console.error(`Unknown preset '${presetName}'. Available: ${Object.keys(presets).join(", ")}`);
  process.exit(1);
}

const workflow = {
  // Load zImageTurbo model (model only — no embedded CLIP/VAE)
  "1": {
    inputs: { ckpt_name: checkpoint },
    class_type: "CheckpointLoaderSimple"
  },
  // Load CLIP separately (Qwen 3 4B, lumina2 type required by zImageTurbo)
  "2": {
    inputs: { clip_name: "qwen_3_4b_fp8_mixed.safetensors", type: "lumina2" },
    class_type: "CLIPLoader"
  },
  // Load VAE separately
  "3": {
    inputs: { vae_name: "ae_zim.safetensors" },
    class_type: "VAELoader"
  },
  // Wrap model with flow-based sampling (shift=3)
  "4": {
    inputs: { model: ["1", 0], shift: 3.0 },
    class_type: "ModelSamplingAuraFlow"
  },
  // Positive prompt
  "5": {
    inputs: { text: preset.positive, clip: ["2", 0] },
    class_type: "CLIPTextEncode"
  },
  // Negative prompt
  "6": {
    inputs: { text: preset.negative, clip: ["2", 0] },
    class_type: "CLIPTextEncode"
  },
  // Latent image
  "7": {
    inputs: { width: size, height: size, batch_size: 1 },
    class_type: "EmptyLatentImage"
  },
  // KSampler: euler_flow + zimage_turbo scheduler
  "8": {
    inputs: {
      seed, steps, cfg,
      sampler_name: "euler_flow",
      scheduler: "zimage_turbo",
      denoise: 1.0,
      model: ["4", 0],
      positive: ["5", 0],
      negative: ["6", 0],
      latent_image: ["7", 0]
    },
    class_type: "KSampler"
  },
  // VAE Decode using separate VAE
  "9": {
    inputs: { samples: ["8", 0], vae: ["3", 0] },
    class_type: "VAEDecode"
  },
  // Save output
  "10": {
    inputs: { filename_prefix: "puffin_texture", images: ["9", 0] },
    class_type: "SaveImage"
  }
};

async function queuePrompt() {
  const res = await fetch(`${endpoint}/prompt`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ prompt: workflow })
  });
  if (!res.ok) throw new Error(`Queue failed: ${res.status}`);
  return res.json();
}

async function waitForImage(promptId) {
  const started = Date.now();
  const maxPolls = Math.max(1, timeoutSec);
  for (let i = 0; i < maxPolls; i++) {
    const res = await fetch(`${endpoint}/history/${promptId}`);
    if (res.ok) {
      const data = await res.json();
      const item = data[promptId];
      const images = item?.outputs?.["10"]?.images;
      if (images?.length) return images[0];
    }
    if ((i + 1) % 10 === 0) {
      const elapsed = Math.floor((Date.now() - started) / 1000);
      console.log(`  waiting... ${elapsed}s`);
    }
    await new Promise((r) => setTimeout(r, 1000));
  }
  throw new Error(`Timed out waiting for image result from ComfyUI after ${timeoutSec}s`);
}

async function downloadImage(meta) {
  const params = new URLSearchParams({
    filename: meta.filename,
    subfolder: meta.subfolder || "",
    type: meta.type || "output"
  });
  const res = await fetch(`${endpoint}/view?${params.toString()}`);
  if (!res.ok) throw new Error(`Download failed: ${res.status}`);
  return Buffer.from(await res.arrayBuffer());
}

async function main() {
  console.log(`Generating '${presetName}' texture via ${endpoint}`);
  const { prompt_id: promptId } = await queuePrompt();
  if (!promptId) throw new Error("No prompt_id returned by ComfyUI");

  const meta = await waitForImage(promptId);
  const bytes = await downloadImage(meta);

  const outDir = path.resolve("img", "generated");
  await fs.mkdir(outDir, { recursive: true });
  const outPath = path.join(outDir, outputName);
  await fs.writeFile(outPath, bytes);

  console.log(`Saved texture: ${outPath}`);
  console.log(`seed=${seed} size=${size} steps=${steps} cfg=${cfg} timeout=${timeoutSec}s preset=${presetName}`);
}

main().catch((err) => {
  console.error(err.message || err);
  process.exit(1);
});

File: scripts/generate-texture.py
#!/usr/bin/env python3
"""
Generate game textures using ComfyUI.
Supports water, ice, rock, lava, grass presets.
"""
import json
import argparse
import time
import urllib.request
import urllib.error
import os
from pathlib import Path

ENDPOINT = "http://127.0.0.1:8188"
PRESET_PATH = Path(__file__).parent / "texture-presets.json"

# Load presets
with open(PRESET_PATH) as f:
    PRESETS = json.load(f)

def load_presets():
    """Load texture presets from JSON file."""
    with open(PRESET_PATH) as f:
        return json.load(f)

def build_workflow(positive_prompt, negative_prompt, size, steps, cfg, seed, checkpoint):
    """Build a ComfyUI workflow for texture generation."""
    return {
        "1": {
            "inputs": {"ckpt_name": checkpoint},
            "class_type": "CheckpointLoaderSimple"
        },
        "2": {
            "inputs": {"text": positive_prompt, "clip": ["1", 1]},
            "class_type": "CLIPTextEncode"
        },
        "3": {
            "inputs": {"text": negative_prompt, "clip": ["1", 1]},
            "class_type": "CLIPTextEncode"  
        },
        "4": {
            "inputs": {"width": size, "height": size, "batch_size": 1},
            "class_type": "EmptyLatentImage"
        },
        "5": {
            "inputs": {
                "seed": seed,
                "steps": steps,
                "cfg": cfg,
                "sampler_name": "euler",
                "scheduler": "normal",
                "denoise": 1.0,
                "model": ["1", 0],
                "positive": ["2", 0],
                "negative": ["3", 0],
                "latent_image": ["4", 0]
            },
            "class_type": "KSampler"
        },
        "6": {
            "inputs": {"samples": ["5", 0], "vae": ["1", 2]},
            "class_type": "VAEDecode"
        },
        "7": {
            "inputs": {"filename_prefix": "puffin_texture", "images": ["6", 0]},
            "class_type": "SaveImage"
        }
    }

def post_prompt(workflow):
    """Submit a workflow to ComfyUI and return prompt_id."""
    url = f"{ENDPOINT}/prompt"
    data = json.dumps({"prompt": workflow}).encode('utf-8')
    req = urllib.request.Request(url, data=data, headers={"Content-Type": "application/json"}, method="POST")
    
    try:
        with urllib.request.urlopen(req) as response:
            result = json.loads(response.read().decode())
            if response.status == 200:
                return result.get("prompt_id")
            else:
                print(f"Error: HTTP {response.status}")
                print(result)
                return None
    except urllib.error.HTTPError as e:
        error_data = json.loads(e.read().decode())
        print(f"HTTP Error {e.code}: {error_data}")
        return None
    except Exception as e:
        print(f"Error submitting prompt: {e}")
        return None

def wait_for_result(prompt_id, timeout=240):
    """Poll ComfyUI for generation results."""
    url = f"{ENDPOINT}/history/{prompt_id}"
    start_time = time.time()
    
    while time.time() - start_time < timeout:
        try:
            with urllib.request.urlopen(url) as response:
                data = json.loads(response.read().decode())
                if prompt_id in data:
                    item = data[prompt_id]
                    # Check if generation is complete
                    if "outputs" in item and "7" in item["outputs"]:
                        images = item["outputs"]["7"].get("images", [])
                        if images:
                            return images[0]
            print(f"Waiting... ({int(time.time() - start_time)}s)")
        except:
            pass
        time.sleep(1)
    
    return None

def download_image(image_meta):
    """Download generated image from ComfyUI."""
    filename = image_meta["filename"]
    subfolder = image_meta.get("subfolder", "")
    
    params = f"filename={filename}&subfolder={subfolder}&type=output"
    url = f"{ENDPOINT}/view?{params}"
    
    try:
        with urllib.request.urlopen(url) as response:
            return response.read()
    except Exception as e:
        print(f"Error downloading image: {e}")
        return None

def main():
    parser = argparse.ArgumentParser(description="Generate game textures using ComfyUI")
    parser.add_argument("--preset", default="water", help="Texture preset (water, ice, rock, lava, grass)")
    parser.add_argument("--size", type=int, default=256, help="Image size in pixels")
    parser.add_argument("--steps", type=int, default=15, help="Generation steps")
    parser.add_argument("--cfg", type=float, default=6.5, help="CFG scale")
    parser.add_argument("--seed", type=int, default=None, help="Random seed")
    parser.add_argument("--ckpt", default="zImageTurbo_turbo.safetensors", help="Checkpoint model")
    parser.add_argument("--out", default=None, help="Output filename")
    
    args = parser.parse_args()
    
    # Validate preset
    if args.preset not in PRESETS:
        print(f"Unknown preset '{args.preset}'. Available: {', '.join(PRESETS.keys())}")
        return 1
    
    preset = PRESETS[args.preset]
    seed = args.seed if args.seed is not None else int(time.time() * 1000) % 2147483647
    output_name = args.out or f"{args.preset}_{int(time.time())}.png"
    
    print(f"Generating '{args.preset}' texture...")
    print(f"  Size: {args.size}x{args.size}, Steps: {args.steps}, CFG: {args.cfg}, Seed: {seed}")
    
    # Build and submit workflow
    workflow = build_workflow(
        preset["positive"],
        preset["negative"],
        args.size,
        args.steps,
        args.cfg,
        seed,
        args.ckpt
    )
    
    prompt_id = post_prompt(workflow)
    if not prompt_id:
        print("Failed to submit prompt to ComfyUI")
        return 1
    
    print(f"Prompt queued: {prompt_id}")
    
    # Wait for result
    image_meta = wait_for_result(prompt_id)
    if not image_meta:
        print("Timeout waiting for generation result")
        return 1
    
    # Download image
    image_data = download_image(image_meta)
    if not image_data:
        print("Failed to download generated image")
        return 1
    
    # Save to file
    out_dir = Path("img") / "generated"
    out_dir.mkdir(parents=True, exist_ok=True)
    out_path = out_dir / output_name
    
    with open(out_path, "wb") as f:
        f.write(image_data)
    
    print(f"✓ Saved texture: {out_path}")
    return 0

if __name__ == "__main__":
    exit(main())
